pkg/engine/pollute.go

```diff
@@ -0,0 +1,155 @@
+/*
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package engine
+
+import (
+	"bytes"
+	"fmt"
+	"os"
+
+	"k8s.io/kubernetes/pkg/api"
+	"k8s.io/kubernetes/pkg/api/meta"
+	"k8s.io/kubernetes/pkg/api/v1"
+	app "k8s.io/kubernetes/pkg/api/v1"
+	"k8s.io/kubernetes/pkg/apimachinery/announced"
+	"k8s.io/kubernetes/pkg/apimachinery/registered"
+	_ "k8s.io/kubernetes/pkg/apis/apps/install"
+	apps "k8s.io/kubernetes/pkg/apis/apps/v1beta1"
+	_ "k8s.io/kubernetes/pkg/apis/authentication/install"
+	_ "k8s.io/kubernetes/pkg/apis/authorization/install"
+	_ "k8s.io/kubernetes/pkg/apis/autoscaling/install"
+	_ "k8s.io/kubernetes/pkg/apis/batch/install"
+	batch "k8s.io/kubernetes/pkg/apis/batch/v1"
+	_ "k8s.io/kubernetes/pkg/apis/certificates/install"
+	_ "k8s.io/kubernetes/pkg/apis/componentconfig/install"
+	_ "k8s.io/kubernetes/pkg/apis/extensions/install"
+	extensions "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	_ "k8s.io/kubernetes/pkg/apis/imagepolicy/install"
+	_ "k8s.io/kubernetes/pkg/apis/policy/install"
+	_ "k8s.io/kubernetes/pkg/apis/rbac/install"
+	_ "k8s.io/kubernetes/pkg/apis/storage/install"
+	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/runtime/serializer/json"
+)
+
+const (
+	// defaultPathKey is the key of chart path
+	defaultPathKey      = "helm.sh/pollutant"
+	defaultNamespaceKey = "helm.sh/namespace"
+	defaultReleaseKey   = "helm.sh/release"
+	defaultRevisionKey  = "helm.sh/revision"
+)
+
+var (
+	// defaultSerializer is a codec and used for encoding and decoding kubernetes resources
+	defaultSerializer *json.Serializer = nil
+)
+
+func init() {
+	// create default serializer
+	schema := runtime.NewScheme()
+	errs := []error{api.AddToScheme(schema), v1.AddToScheme(schema)}
+	for _, err := range errs {
+		if err != nil {
+			panic(err)
+		}
+	}
+	announced.DefaultGroupFactoryRegistry.RegisterAndEnableAll(registered.NewOrDie(os.Getenv("KUBE_API_VERSIONS")), schema)
+	defaultSerializer = json.NewYAMLSerializer(json.DefaultMetaFactory, schema, schema)
+}
+
+// pollute adds pollutant to resource annotations. If resource is not a valid kubernetes
+// resource, it does nothing and returns original resource.
+func pollute(resource string, r *renderable) string {
```

Comment:
How's pollution used?  I guess in kubernetes-admin where you query use annotation?